service/orchestrator/certificates.go

```diff
@@ -0,0 +1,152 @@
+	"golang.org/x/exp/slices"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+// CreateCertificate implements method for creating a new certificate
+func (svc *Service) CreateCertificate(ctx context.Context, req *orchestrator.CreateCertificateRequest) (
+	res *orchestrator.Certificate, err error) {
+
+	// Validate request
+	if err = service.ValidateRequest(req); err != nil {
+		err = status.Error(codes.InvalidArgument, err.Error())
+		return
+	}
+
+	// Check if client is allowed to access the corresponding cloud service (targeted in the certificate)
+	if !svc.authz.CheckAccess(ctx, service.AccessCreate, req) {
+		err = service.ErrPermissionDenied
+		return
+	}
+
+	// Persist the new certificate in our database
+	err = svc.storage.Create(req.Certificate)
+	if err != nil {
+		err = status.Errorf(codes.Internal, "could not add certificate to the database: %v", err)
+		return
+	}
+
+	logging.LogRequest(log, logrus.DebugLevel, logging.Create, req)
+
+	// Return certificate
+	res = req.Certificate
+	return
+}
+
+// GetCertificate implements method for getting a certificate, e.g. to show its state in the UI
+func (svc *Service) GetCertificate(ctx context.Context, req *orchestrator.GetCertificateRequest) (
+	res *orchestrator.Certificate, err error) {
+
+	// Validate request
+	if err = service.ValidateRequest(req); err != nil {
+		err = status.Error(codes.InvalidArgument, err.Error())
+		return
+	}
+
+	res = new(orchestrator.Certificate)
+	err = svc.storage.Get(res, "Id = ?", req.CertificateId)
+	if errors.Is(err, persistence.ErrRecordNotFound) {
+		return nil, status.Errorf(codes.NotFound, "certificate not found")
+	} else if err != nil {
+		return nil, status.Errorf(codes.Internal, "database error: %v", err)
+	}
+
+	// Check if client is allowed to access the corresponding cloud service (targeted in the certificate)
+	areAllAllowed, allowedCloudServices := svc.authz.AllowedCloudServices(ctx)
+	if !areAllAllowed && !slices.Contains(allowedCloudServices, res.CloudServiceId) {
+		// Important to nil the response since it is set already
+		return nil, service.ErrPermissionDenied
+	}
+
+	return
+}
+
+// ListCertificates implements method for getting a certificate, e.g. to show its state in the UI
+func (svc *Service) ListCertificates(ctx context.Context, req *orchestrator.ListCertificatesRequest) (res *orchestrator.ListCertificatesResponse, err error) {
+	// Validate request
+	err = service.ValidateRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	res = new(orchestrator.ListCertificatesResponse)
+
+	res.Certificates, res.NextPageToken, err = service.PaginateStorage[*orchestrator.Certificate](req, svc.storage,
+		service.DefaultPaginationOpts)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "could not paginate results: %v", err)
+	}
+
+	// Only list certificates user is authorized to see
+	var certificateSlice []*orchestrator.Certificate
```

Comment:
Instead of doing that, you should append an additional query parameter, that checks if the cloud service ID of the certificate is in a list of allowed services. See other List methods for examples.

Reply:
Ahh. You mean, first get all allowed cloud services and insert this into a query?

Reply:
Yes